### PR TITLE
Value expectation and 1st order CKY

### PR DESCRIPTION
**Pull request by teffland**
Changes are:
* Add 1st order cky implementation with suggested updates
* Add value-expectation semiring and test
* Make marginal computation work on potentials w/o gradients
* Minor clarity/consistency changes to semirings
* Some documentation for _Struct

### CODE REVIEW

**Comment by srush** [comment]
Hmm, for some reason the tests are not running on this. Trying to figure out why.

File: tests/test_distributions.py [hunk 1]
```diff
@@ -76,6 +76,18 @@ def test_simple(data, seed):
...
+    val_func = torch.rand(*vals.shape, 10)
+    E_val = dist.expected_value(val_func)
+    struct_vals = (
+        edges.unsqueeze(-1)
+        .mul(val_func.unsqueeze(0))
+        .reshape(*edges.shape[:2], -1, val_func.shape[-1])
+        .sum(2)
+    )
+    assert torch.isclose(
+        E_val, log_probs.exp().unsqueeze(-1).mul(struct_vals).sum(0)
```
Comment (srush) [inline tests/test_distributions.py:88]: Just curious. Why not just make this the implementation of expected value? It seems just as good and perhaps more efficient.y

File: torch_struct/distributions.py [hunk 2]
```diff
@@ -179,11 +181,49 @@ def marginals(self):
...
+        "Compute the total number of structures in the CRF support set."
         ones = torch.ones_like(self.log_potentials)
         ones[self.log_potentials.eq(-float("inf"))] = 0
         return self._struct(StdSemiring).sum(ones, self.lengths)
 
+    def expected_value(self, values):
+        """
+        Compute expectated value for distribution :math:`E_z[f(z)]` where f decomposes additively over the factors of p_z.
+
+        Params:
```
Comment (srush) [inline torch_struct/distributions.py:193]: This should be "Parameters:"

File: torch_struct/distributions.py [hunk 3]
```diff
@@ -179,11 +181,49 @@ def marginals(self):
...
+          * values (*batch_shape x *event_shape, *value_shape): torch.FloatTensor that assigns a value to each part
```
Comment (srush) [inline torch_struct/distributions.py:194]: Let's put the types in the first parens, and use :class:`torch.FloatTensor`

File: torch_struct/distributions.py [hunk 4]
```diff
@@ -204,26 +244,30 @@ def partition(self):
...
-        nsamples = sample_shape[0]
+        if type(sample_shape) == int:
+            nsamples = sample_shape
+        else:
+            assert len(sample_shape) == 1
+            nsamples = sample_shape[0]
         samples = []
         for k in range(nsamples):
-            if k % 10 == 0:
+            if k % batch_size == 0:
```
Comment (srush) [inline torch_struct/distributions.py:265]: Oh yeah, sorry this is my fault. 10 is a global constant. Let's put it on MultiSampledSemiring.

File: torch_struct/distributions.py [hunk 5]
```diff
@@ -411,6 +455,32 @@ class TreeCRF(StructDistribution):
...
+
+    Parameters:
+        log_potentials (tensor) : event_shape (*N x N x N x NT x NT x NT*), e.g.
+                                    :math:`\phi(i, j, k, A_i^j \rightarrow B_i^k C_{k+1}^j)`
+        lengths (long tensor) : batch shape integers for length masking.
+
+    Implementation uses width-batched, forward-pass only
+
+    * Parallel Time: :math:`O(N)` parallel merges.
+    * Forward Memory: :math:`O(N^2)`
```
Comment (srush) [inline torch_struct/distributions.py:477]: This can't be right... isn't the event shape O(N^3) alone?

File: torch_struct/full_cky_crf.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import torch
+from .helpers import _Struct, Chart
+from tqdm import tqdm
```
Comment (srush) [inline torch_struct/full_cky_crf.py:3]: Be sure to run `python setup.py style` to run flake8 . It will catch these errors.

File: torch_struct/helpers.py [hunk 7]
```diff
@@ -24,15 +24,44 @@ def __setitem__(self, ind, new):
...
+        """Implement computation equivalent to the computing log partition constant logZ (if self.semiring == `_BaseSemiring`).
+
+        Parameters:
+          scores (torch.FloatTensor) : log potential scores for each factor of the model. Shape (* x batch size x *event_shape )
+          lengths (torch.LongTensor) : = None, lengths of batch padded examples. Shape = ( * x batch size )
+          force_grad: bool = False
+
+        Returns:
+          v (torch.Tensor) : the resulting output of the dynammic program
+          edges (List[torch.Tensor]): the log edge potentials of the model.
```
Comment (srush) [inline torch_struct/helpers.py:46]: changing this to `logpotentials` throughout.

File: torch_struct/helpers.py [hunk 8]
```diff
@@ -24,15 +24,44 @@ def __setitem__(self, ind, new):
...
+                 When `scores` is already in a log_potential format for the distribution (typical), this will be
+                 [scores], as in `Alignment`, `LinearChain`, `SemiMarkov`, `CKY_CRF`.
+                 An exceptional case is the `CKY` struct, which takes log potential parameters from production rules
+                 for a PCFG, which are by definition independent of position in the sequence.
+          charts: Optional[List[Chart]] = None, the charts used in computing the dp. They are needed if we want to run the
```
Comment (srush) [inline torch_struct/helpers.py:51]: Going to remove this for simplicity.

File: torch_struct/helpers.py [hunk 9]
```diff
@@ -92,28 +121,31 @@ def marginals(self, logpotentials, lengths=None, _raw=False):
             marginals: b x (N-1) x C x C table
 
         """
-        v, edges = self.logpartition(logpotentials, lengths=lengths, force_grad=True)
-        if _raw:
-            all_m = []
-            for k in range(v.shape[0]):
-                obj = v[k].sum(dim=0)
-
+        with torch.autograd.enable_grad():  # in case input potentials don't have grads enabled.
```
Comment (srush) [inline torch_struct/helpers.py:124]: Cool

File: torch_struct/semirings/semirings.py [hunk 10]
```diff
@@ -468,22 +469,94 @@ def prod(cls, xs, dim=-1):
...
-        xs[0].fill_(0)
-        xs[1].fill_(0)
+    @classmethod
+    def one_(cls, xs):
+        xs[0].fill_(cls.one[0])
+        xs[1].fill_(cls.one[1])
         return xs
 
 
+def ValueExpectationSemiring(k):
```
Comment (srush) [inline torch_struct/semirings/semirings.py:488]: Are you sure we don't have this already? Could have sworn someone added it.

**Review by srush — CHANGES_REQUESTED** [review]

**Comment by srush** [comment]
Thanks the PR. Lots of nice stuff in here.

File: tests/test_distributions.py (hunk 1 above)
Comment (teffland) [reply tests/test_distributions.py:88]: Sorry, maybe I'm confused but isn't this enumerating over all possible structures explicitly?

File: torch_struct/distributions.py (hunk 5 above)
Comment (teffland) [reply torch_struct/distributions.py:477]: Oops yeah that's from modifying the CKYCRF class

File: torch_struct/semirings/semirings.py (hunk 10 above)
Comment (teffland) [reply torch_struct/semirings/semirings.py:488]: I'm not 100% sure, I looked and hadn't seen it anywhere in master so I went ahead with it. Maybe it's in another branch?

File: tests/test_distributions.py (hunk 1 above)
Comment (srush) [reply tests/test_distributions.py:88]: Oh sorry, my comment is confusing. 

I think a valid way of computing an expectation over any "part-level value" is to first compute the marginals (.marginals()) and then doing an elementwise mul (.mul) and then summing. Doesn't that give you the same thing as the semiring?

Comment (teffland) [reply tests/test_distributions.py:88]: Oh wow, I didn't realize this! I just tested it out and it appears to be more efficient for larger structure sizes. I guess this is due to the fast log semiring implementation? I'll update things to use this approach instead.

Comment (srush) [reply tests/test_distributions.py:88]: Yeah, I think that is right... I haven't thought about this too much, but my guess is that this is just better on GPU hardware since the expectation is batched at the end. But it seems worth understand when this works. I don't think you can compute Entropy this way? (but I might be wrong)

Comment (teffland) [reply tests/test_distributions.py:88]: Makes sense. I also don't think entropy can be done this way -- I just tested it out and the results didn't match the semiring. I will switch to this implementation in the latest commit and get rid of the value semiring.


Fwiw I ran a quick speed comparison you might be interested in:

```
B, N, C = 4, 200, 10

phis = torch.randn(B,N,C,C).cuda()
vals = torch.randn(B,N,C,C,10).cuda()
```

### Results from running w/ genbmm
```
%%timeit
LinearChainCRF(phis).expected_value(vals)
>>> 100 loops, best of 3: 6.34 ms per loop

%%timeit
LinearChainCRF(phis).marginals.unsqueeze(-1).mul(vals).reshape(B,-1,vals.shape[-1]).sum(1)
>>> 100 loops, best of 3: 5.64 ms per loop
```

### Results from running w/o genbmm

```
%%timeit
LinearChainCRF(phis).expected_value(vals)
>>> 100 loops, best of 3: 9.67 ms per loop

%%timeit
LinearChainCRF(phis).marginals.unsqueeze(-1).mul(vals).reshape(B,-1,vals.shape[-1]).sum(1)
>>> 100 loops, best of 3: 8.83 ms per loop
```

**Comment by teffland** [comment]
Quick dev question: when I try running `python setup.py test` or `python setup.py style`, I get `error: invalid command 'tests'` and `error: option --ignore not recognized`, respectively. I can run the command on their own, but I'm just wondering if I'm doing something wrong?

**Comment by srush** [comment]
Interesting, yeah not sure how to run those automatically, I will look into
it.
